crates/fs-lock/src/lib.rs

```diff
@@ -24,6 +24,16 @@ impl FileLock {
         Ok(Self(file))
     }
 
+    /// Try to take an exclusive lock on a [`File`].
+    ///
+    /// Note that this operation is blocking, and should not be called in async contexts.
+    pub fn new_try_exclusive(file: File) -> std::result::Result<Self, File> {
```

Comment:
```suggestion
    pub fn new_try_exclusive(file: File) -> std::result::Result<Self, (io::Error, File)> {
```

Reply:
You could change `new_exclusive` and `new_share` to return `io::Result` so you can use `Result` instead of `std::result::Result` here

Reply:
Is returning this `io::Error` useful? Can these operations fail for reasons other than just failing to get the lock?

I guess they probably can...

Reply:
@NobodyXu I did `(File, io::Error)`, seems to me like a more idiomatic order.

Reply:
@NobodyXu There's an API problem here that the user has no good way to tell if it was locking error or some other `io::Result`.

We could turn it into `Result<Self, (Option<io::Error>, File)>` where `Err(None)` means "locking failed". It's a bit of an ugly API, but I don't expect many people to use it, I can live with it, and it does a job.

Reply:
That sounds good to me.

Reply:
According to https://github.com/search?q=repo%3Aal8n%2Ffs4-rs%20lock_error&type=code seems like comparing by `raw_os_error()` is the way to go. @NobodyXu

Reply:
Yeah that will be ok

Reply:
@NobodyXu https://doc.rust-lang.org/std/io/struct.Error.html#method.raw_os_error common method, or I misunderstand you.

Reply:
Well, the https://doc.rust-lang.org/std/io/type.RawOsError.html is said to be unstable, but the code compiles with stable Rust, and I would expect this type to always implement equality, otherwise what's really the point of it.

Reply:
> https://doc.rust-lang.org/std/io/type.RawOsError.html is said to be unstable

It's just a type alias, in reality it always returns i32

Reply:
Yes